lib/safe_macros.c

```diff
@@ -210,7 +210,12 @@ int safe_open(const char *file, const int lineno, void (*cleanup_fn) (void),
 	mode_t mode;
 
 	va_start(ap, oflags);
-	mode = va_arg(ap, mode_t);
+
+	if (sizeof(mode_t) < sizeof(int))
+		mode = va_arg(ap, int);
+	else
+		mode = va_arg(ap, mode_t);
+
 	va_end(ap);
 
 	rval = open(pathname, oflags, mode);
```

Comment:
You are right we get the mode in an integer sized storage. Looking into the glibc internals the mode_t is declared as 32bit integer, so this is likely the reason why this works fine on Linux with glibc. Btw 16bit integer for mode_t sounds like a useless optimization to me...

On the other hand I doubt that there is a system where `sizeof(mode_t) > sizeof(int)` so I'm tempted to simplify this to just `mode = va_arg(ap, int)`, what do you think?